numpy/array_api/_array_object.py

```diff
@@ -124,6 +125,9 @@ def __array__(self, dtype: None | np.dtype[Any] = None) -> npt.NDArray[Any]:
     # spec in places where it either deviates from or is more strict than
     # NumPy behavior
 
+    # NOTE: no valid type annotation possible.  E.g `Union[Array,
+    # NotImplemented]` is forbidden, see https://github.com/python/mypy/issues/363
+    # Maybe change returned object to `Literal['NotImplemented']`?
     def _check_allowed_dtypes(self, other, dtype_category, op):
```

Comment:
```suggestion
    def _check_allowed_dtypes(self, other: bool | int | float | Array, dtype_category: str, op: str) -> Array:
```
[`NotImplemented`](https://github.com/python/typeshed/blob/b22f29bec136649d736b70d67abbe079c672b044/stdlib/builtins.pyi#L972-L978) is considered equivalent to `Any` as far as static type checkers are concerned.
This was done so that, for example, methods like `__eq__` can simply return `bool` instead of the more verbose `bool | Literal[NotImplemented]`.

For now, I'd suggest to just ignore `NotImplemented` as far as return types concerned.

Reply:
Makes sense, thanks. Tested locally, and this works just fine with Mypy 0.930